### PR TITLE
Add mermaidjs as format output for pyreverse

### PR DESCRIPTION
**Pull request by sgheppy**
This pull request add the [mermaidjs](https://mermaid-js.github.io/mermaid/#/classDiagram) output format (as text like plantuml)

For the packages graph the classDiagram type of diagram is used for mermaidjs, because is the nearest type of graph to the package diagram for this tool.

### CODE REVIEW

**Comment by DudeNr33** [comment]
Very nice, did not know of mermaidjs before. Can you add tests as well please? You can find the existing tests under ``tests/pyreverse/test_writer.py``. Will review the rest of the PR as soon as the tests are added!

File: pylint/pyreverse/mermaidjs_printer.py
Comment (DudeNr33) [inline pylint/pyreverse/mermaidjs_printer.py:1]: I think your name is the one that should appear here! 😊

File: ChangeLog
Comment (DudeNr33) [inline ChangeLog:14]: That should probably be mermaidjs instead of PlantUML.

File: doc/whatsnew/2.12.rst
Comment (DudeNr33) [inline doc/whatsnew/2.12.rst:18]: This should be moved into the "Other changes" section. There also already is a bullet point for your last fix, so the "Fix ``install graphviz`` message ..." can be removed.

**Review by DudeNr33 — COMMENTED** [review]
I already left some remarks in the other files, but I will wait for the tests before I start reviewing the new ``MermaidJSPrinter``.

File: pylint/pyreverse/mermaidjs_printer.py
Comment (Pierre-Sassoulas) [reply pylint/pyreverse/mermaidjs_printer.py:1]: You can remove the copyright it will be automatically generated before release :)

**Comment by DudeNr33** [comment]
I was thinking if the final output file should already contain the minimal HTML boilerplate necessary to directly open the result in a browser. I think being able to directly view the diagram without having to install Graphviz or PlantUML is one of the advantages of the mermaidjs format. 
It does not require much to get it to work:
```html
<html>

<body>
    <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
    <div class="mermaid">
        graph TD
        A[Party] -->|Get money| B(Go shopping!!)
    </div>
</body>

</html>
```
It would be even possible to omit the ``html`` tag, but I think it's cleaner with it.

@Pierre-Sassoulas what are your thoughts on this?

**Comment by Pierre-Sassoulas** [comment]
> I was thinking if the final output file should already contain the minimal HTML boilerplate necessary to directly open the result in a browser.

It depends if you want to use it directly or feed it in another program to be included elsewhere. I think both use are valid, the default could be viewable directly as html, and there could be an option to only output the code without boilerplate ? What do you think @sgheppy ?

**Comment by sgheppy** [comment]
I think is a good idea.
maybe could be an alternative output like png for graphviz.
So if the requested output format is mmd the simple mermaidjs format is returned.
On the other hand, if the requested output is html, the minimal HTML boilerplate is added.

What do you think @Pierre-Sassoulas?

**Comment by Pierre-Sassoulas** [comment]
>  if the requested output is html, the minimal HTML boilerplate is added.

Amazing design idea ! This is very intuitive.

**Comment by DudeNr33** [comment]
I like that idea as well. 
Thanks for adding the tests!
It would be also nice if you could add the new output options to the documentation (``doc/additional_commands/index.rst``).

**Comment by sgheppy** [comment]
the precommit check, edit the html data check file. How i can avoid that?

**Comment by DudeNr33** [comment]
It's best practice to also [install pre-commit in your own development environment](https://pylint.pycqa.org/en/latest/development_guide/contribute.html?highlight=pre-commit#repository).

As you already noticed, in your case that is not enough. You need to exclude the directory for the output files from the pre-commit hook that alters them (my guess would be ``prettier``, but you'll see that when you run it locally). This can be done in ``.pre-commit-config.yaml``.

File: .pre-commit-config.yaml
Comment (DudeNr33) [inline .pre-commit-config.yaml:63]: This is not the correct place for the exclude pattern. You can add different excludes for each hook, and my guess is that the HTML reformatting is performed by ``prettier`` (the last hook in the file).
If you install ``pre-commit`` on your own machine, you can see which hook changes which files before you commit.

You can install it by following the pylint docs:

> * install pre-commit using ``pip install pre-commit``
> * then run ``pre-commit install`` in the pylint root directory to enable the git hooks.

Comment (sgheppy) [reply .pre-commit-config.yaml]: thank you for your suggestions, i really appreciate those. I'm using gitpod to edit, and seems that pre-commit has an issue in that behavior. After a (not so) quick and a (really) dirty fix i used pre-commit in this environment.

I hope it's ok now.

Thank you for your support

File: .pre-commit-config.yaml
Comment (Pierre-Sassoulas) [inline .pre-commit-config.yaml:9]: Could we target specifically the html for pyreverse's tests here and for prettier, please ?
```suggestion
        exclude: "tests/functional/t/trailing_whitespaces.py|tests/pyreverse/data/*.html"
```

Comment (sgheppy) [reply .pre-commit-config.yaml:9]: Sure

File: doc/whatsnew/2.12.rst
Comment (DudeNr33) [inline doc/whatsnew/2.12.rst:19]: Can you please move this to ``doc/whatsnew/2.13.rst``, and put it below "Extensions" instead.

File: pylint/pyreverse/mermaidjs_printer.py
Comment (DudeNr33) [inline pylint/pyreverse/mermaidjs_printer.py:105]: A few more ``_inc_indent()`` would be nice so that the mermaid code has a proper indentation under the ``div``.

Comment (sgheppy) [reply pylint/pyreverse/mermaidjs_printer.py:106]: added indentation

File: doc/whatsnew/2.12.rst
Comment (DudeNr33) [inline doc/whatsnew/2.12.rst:89]: This still needs to be moved to ``doc/whatsnew/2.13.rst`` I think, as Pierre set the milestone to 2.13.0 for this PR.
@Pierre-Sassoulas or do you plan to include it in an upcoming 2.12.x release?

File: ChangeLog
Comment (DudeNr33) [reply ChangeLog:43]: This still needs to be changed to MermaidJS, and moved to the ``2.13`` section.

File: doc/whatsnew/2.12.rst
Comment (sgheppy) [reply doc/whatsnew/2.12.rst]: Sorry, it was an oversight. Thank you for your reviews.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, thanks a lot @sgheppy !  @DudeNr33 I'll let you merge when it's ok for you :)

**Comment by Pierre-Sassoulas** [comment]
Sorry I forgot that we need when to release pylint 2.12.2 first, I'm adding the blocked label.

**Comment by sgheppy** [comment]
> LGTM, thanks a lot @sgheppy !  @DudeNr33 I'll let you merge when it's ok for you :) 

Thank you all for your support!

**Comment by sgheppy** [comment]
> Sorry I forgot that we need when to release pylint 2.12.2 first, I'm adding the blocked label. 

Do I have to do something about it?

**Comment by Pierre-Sassoulas** [comment]
> Do I have to do something about it?

No, it's maintenance work, we'll merge and rebase if required when it's ready  :)

**Review by DudeNr33 — APPROVED** [review]
No more comments from my side, thanks for incorporating all the feedback @sgheppy!

@Pierre-Sassoulas I'll wait until the ``blocked`` label is removed before I merge.
I think "rebase and merge" would be the best option in this case then?

**Comment by Pierre-Sassoulas** [comment]
> I think "rebase and merge" would be the best option in this case then?

Generally when there is a pre-commit automated commit I squash because it means the author did not clean up its tree. I also consider the number of added / removed line (how messy would it be with only one commit?), and quality of messages for the commit. If there are only quality messages I rebase. If there are unclear message and a lot of added OR a lot of removed it's not an issue to have a big commit so I squash. For a lot of added AND a lot of remove with bad commit message I usually ask for a rebase/cleanup from the author. (Right now I think we could squash because  a single commit with almost only addition make sense). I'm just sharing my thought process, to get the better end result it's not a hard "rule" 😄

**Comment by DudeNr33** [comment]
Thanks for the guideline!

**Comment by DanielNoord** [comment]
After this merge `pytest -k test_writer` leaves test files in the repository. This is a bit annoying when running the full test suite. I haven't looked at the code myself yet, but would anyone be able to do a follow up PR so the files created during testing are removed after testing is done?

**Comment by sgheppy** [comment]
> After this merge `pytest -k test_writer` leaves test files in the repository. This is a bit annoying when running the full test suite. I haven't looked at the code myself yet, but would anyone be able to do a follow up PR so the files created during testing are removed after testing is done?

I'll do something for this.
Thank you for your suggestion.

**Comment by DudeNr33** [comment]
I missed that in the review. Currently the ``_setup`` function used by the fixtures takes care of cleaning up, but you need to add the ``MMD_FILES`` and ``HTML_FILES`` lists explicitly in the ``for`` loop after the ``yield``.

**Comment by sgheppy** [comment]
> I missed that in the review. Currently the `_setup` function in `tests/pyreverse/test_writer.py` used by the fixtures takes care of cleaning up, but you need to add the `MMD_FILES` and `HTML_FILES` lists explicitly in the `for` loop after the `yield`.

suggested edits implemented in #5476

**Pull request merged**
